runtime/src/accounts_cache.rs

```diff
@@ -47,11 +48,11 @@ impl SlotCacheInner {
         );
     }
 
-    pub fn insert(
+    pub fn insert<T: GetHash>(
         &self,
         pubkey: &Pubkey,
         account: AccountSharedData,
-        hash: Option<Hash>,
+        hash: Option<T>,
```

Comment:
I think the more Rusty way to accept both `T` and `&T` is to pass `impl std::borrow::Borrow<T>`, i.e. `hash: Option<impl Borrow<Hash>>`,

Reply:
love it!

Reply:
`AsRef<Hash>` should work as well

Reply:
is one preferred over the other? For performance? For rustiness? Also, if I have access to a value that I can move, do either of these accomplish that or do they always copy somehow? There are subtleties I don't understand yet.

Reply:
There is a subtle difference between `AsRef` and `Borrow`. From the book: 

```
Choose Borrow when you want to abstract over different kinds of borrowing, or when you’re building a data structure that treats owned and borrowed values in equivalent ways, such as hashing and comparison.

Choose AsRef when you want to convert something to a reference directly, and you’re writing generic code.
```
Also from the docs: https://doc.rust-lang.org/nightly/std/convert/trait.AsRef.html

```
Unlike AsRef, Borrow has a blanket impl for any T, and can be used to accept either a reference or a value.
Borrow also requires that Hash, Eq and Ord for borrowed value are equivalent to those of the owned value. For this reason, if you want to borrow only a single field of a struct you can implement AsRef, but not Borrow.
```

I believe `Borrow` is more appropriate here

Reply:
I implemented borrow. It does what I want and indications are it generates the code I want.